store/sqlstore/channel_store.go

```diff
@@ -4097,3 +4097,159 @@ func (s SqlChannelStore) GetTeamForChannel(channelID string) (*model.Team, error
 	}
 	return &team, nil
 }
+
+// GetTopChannelsForTeamSince returns the filtered post counts of the following Channels sets:
+// a) those that are private channels in the given user's membership graph on the given team, and
+// b) those that are public channels in the given team.
+func (s SqlChannelStore) GetTopChannelsForTeamSince(teamID string, userID string, since int64, offset int, limit int) (*model.TopChannelList, error) {
+	channels := make([]*model.TopChannel, 0)
+	var args []interface{}
+	postgresPropQuery := `AND (Posts.Props ->> 'from_bot' IS NULL OR Posts.Props ->> 'from_bot' = 'false')`
+	mySqlPropsQuery := `AND (JSON_EXTRACT(Posts.Props, '$.from_bot') IS NULL OR JSON_EXTRACT(Posts.Props, '$.from_bot') = 'false')`
+
+	query := `
+		SELECT
+			ID,
+			Type,
+			DisplayName,
+			Name,
+			TeamID,
+			MessageCount
+		FROM
+			((SELECT
+				Posts.ChannelId AS ID,
+				Channels.Type AS Type,
+				Channels.DisplayName AS DisplayName,
+				Channels.Name AS Name,
+				Channels.TeamId AS TeamID,
+				count(Posts.Id) AS MessageCount
+			FROM
+				Posts
+				LEFT JOIN Channels on Posts.ChannelId = Channels.Id
+			WHERE
+				Posts.DeleteAt = 0
+				AND Posts.CreateAt > ?
+				AND Posts.Type = ''`
+	args = []interface{}{since}
+
+	if s.DriverName() == model.DatabaseDriverMysql {
+		query += mySqlPropsQuery
+	} else if s.DriverName() == model.DatabaseDriverPostgres {
+		query += postgresPropQuery
+	}
+
+	query += `
+				AND Channels.DeleteAt = 0
+				AND Channels.TeamId = ?
+				AND Channels.Type = 'O'
```

Comment:
I have to keep digging, but for some reason if I comment-out any one of these three `where` clauses on the `Channels` table the query time decreases from about 6s to <1s on MySQL in our large test dataset.

Reply:
For some reason putting the `WHERE DeleteAt = 0` in the outer query speeds thing up a lot:

```diff
SELECT
    ID,
    Type,
    DisplayName,
    Name,
    TeamID,
    MessageCount
FROM
    (
        (
            SELECT
                Posts.ChannelId AS ID,
                Channels.Type AS Type,
                Channels.DisplayName AS DisplayName,
                Channels.Name AS Name,
                Channels.TeamId AS TeamID,
                count(Posts.Id) AS MessageCount,
+                Channels.DeleteAt AS DeleteAt
            FROM
                Posts
                LEFT JOIN Channels on Posts.ChannelId = Channels.Id
            WHERE
                Posts.DeleteAt = 0
                AND Posts.CreateAt > 1635996172311
                AND Posts.Type = ''
                AND (
                    JSON_EXTRACT(Posts.Props, '$.from_bot') IS NULL
                    OR JSON_EXTRACT(Posts.Props, '$.from_bot') = 'false'
                )
-                AND Channels.DeleteAt = 0
                AND Channels.TeamId = 'd6z6ei9gmpdyfys5ay7opket8o'
                AND Channels.Type = 'O'
            GROUP BY
                Posts.ChannelId,
                Channels.Type,
                Channels.DisplayName,
                Channels.Name,
                Channels.TeamId,
                Channels.DeleteAt
        )
        UNION
        ALL(
            SELECT
                Posts.ChannelId AS ID,
                Channels.Type AS Type,
                Channels.DisplayName AS DisplayName,
                Channels.Name AS Name,
                Channels.TeamId AS TeamID,
                count(Posts.Id) AS MessageCount,
+                Channels.DeleteAt AS DeleteAt
            FROM
                Posts
                LEFT JOIN Channels on Posts.ChannelId = Channels.Id
                LEFT JOIN ChannelMembers on Posts.ChannelId = ChannelMembers.ChannelId
            WHERE
                Posts.DeleteAt = 0
                AND Posts.CreateAt > 1635996172311
                AND Posts.Type = ''
                AND (
                    JSON_EXTRACT(Posts.Props, '$.from_bot') IS NULL
                    OR JSON_EXTRACT(Posts.Props, '$.from_bot') = 'false'
                )
-                AND Channels.DeleteAt = 0
                AND Channels.TeamId = 'd6z6ei9gmpdyfys5ay7opket8o'
                AND Channels.Type = 'P'
                AND ChannelMembers.UserId = '116ny8fhbb8fbfdeauq73t591o'
            GROUP BY
                Posts.ChannelId,
                Channels.Type,
                Channels.DisplayName,
                Channels.Name,
                Channels.TeamId,
                Channels.DeleteAt
        )
    ) AS A
+WHERE DeleteAt = 0
ORDER BY
    MessageCount DESC,
    Name ASC
LIMIT
    61 OFFSET 0;
```

Reply:
btw, I'm totally OK with optimizing this query (if needed) later and moving forward as-is, assuming logical correctness.